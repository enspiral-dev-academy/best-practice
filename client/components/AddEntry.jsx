import React, { useState } from "react";
import { Link } from "react-router-dom";

import {
  wrappedWith,
  authentication,
  entryContext,
  userContext,
} from "../wrappers";

const formStyle = {
  display: "flex",
  flexDirection: "column",
  maxWidth: "250px",
};

export function AddEntry({ authenticated, user, submitEntry }) {
  const [formData, setFormData] = useState({
    name: "",
    link: "",
    description: "",
  });

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData({ ...formData, [name]: value });
  };

  const handleAdd = () => {
    const authorId = user.id;
    submitEntry(authorId, formData);
  };

  const { name, link, description } = formData;
  return (
    <div data-testid="addentry">
      <h2>Add New Entry</h2>
      {authenticated() ? (
        <div style={formStyle}>
          <label htmlFor="name">Name</label>
          <input id="name" name="name" value={name} onChange={handleChange} />

          <label htmlFor="link">Link</label>
          <input id="link" name="link" value={link} onChange={handleChange} />

          <label htmlFor="description">Description</label>
          <textarea
            id="description"
            name="description"
            value={description}
            onChange={handleChange}
            cols="26"
          />

          <div>
            <button type="button" onClick={handleAdd}>
              Add this entry
            </button>
          </div>
        </div>
      ) : (
        <div>
          You must <Link to="/signin">sign in</Link> to add new entries.
        </div>
      )}
    </div>
  );
}

export default wrappedWith(authentication, entryContext, userContext)(AddEntry);
