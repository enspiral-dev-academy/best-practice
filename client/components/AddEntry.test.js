import React from "react";
import "@testing-library/jest-dom";
import { screen, fireEvent } from "@testing-library/react";
import { renderWithRouter } from "../testing/utils";

import { AddEntry } from "./AddEntry";

test("renders correctly when authenticated", () => {
  const { asFragment } = renderWithRouter(
    <AddEntry authenticated={() => true} />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders correctly when not authenticated", () => {
  const { asFragment } = renderWithRouter(
    <AddEntry authenticated={() => false} />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("submitEntry is called on add entry button click", async () => {
  const submitEntry = jest.fn();
  renderWithRouter(
    <AddEntry
      user={{ id: 1 }}
      submitEntry={submitEntry}
      authenticated={() => true}
    />
  );

  const addButton = await screen.getByRole("button", {
    name: "Add this entry",
  });
  fireEvent.click(addButton);

  expect(submitEntry).toHaveBeenCalled();
});

test("input values update on change", async () => {
  renderWithRouter(<AddEntry authenticated={() => true} />);

  const nameInput = await screen.getByRole("textbox", { name: "Name" });
  const linkInput = await screen.getByRole("textbox", { name: "Link" });
  const descriptionInput = await screen.getByRole("textbox", {
    name: "Description",
  });
  fireEvent.change(nameInput, { target: { value: "test name" } });
  fireEvent.change(linkInput, { target: { value: "test link" } });
  fireEvent.change(descriptionInput, { target: { value: "test description" } });

  expect(nameInput.value).toBe("test name");
  expect(linkInput.value).toBe("test link");
  expect(descriptionInput.value).toBe("test description");
});
