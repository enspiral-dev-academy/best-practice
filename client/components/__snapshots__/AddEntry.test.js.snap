// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when authenticated 1`] = `
<DocumentFragment>
  <div
    data-testid="addentry"
  >
    <h2>
      Add New Entry
    </h2>
    <div
      style="display: flex; flex-direction: column; max-width: 250px;"
    >
      <label
        for="name"
      >
        Name
      </label>
      <input
        id="name"
        name="name"
        value=""
      />
      <label
        for="link"
      >
        Link
      </label>
      <input
        id="link"
        name="link"
        value=""
      />
      <label
        for="description"
      >
        Description
      </label>
      <textarea
        cols="26"
        id="description"
        name="description"
      />
      <div>
        <button
          type="button"
        >
          Add this entry
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly when not authenticated 1`] = `
<DocumentFragment>
  <div
    data-testid="addentry"
  >
    <h2>
      Add New Entry
    </h2>
    <div>
      You must 
      <a
        href="/signin"
      >
        sign in
      </a>
       to add new entries.
    </div>
  </div>
</DocumentFragment>
`;
